cmd/docker-app/cnab.go

```diff
@@ -0,0 +1,185 @@
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/deislabs/duffle/pkg/bundle"
+	"github.com/deislabs/duffle/pkg/credentials"
+	"github.com/deislabs/duffle/pkg/driver"
+	"github.com/deislabs/duffle/pkg/duffle/home"
+	"github.com/deislabs/duffle/pkg/loader"
+	"github.com/docker/app/internal"
+	"github.com/docker/app/internal/packager"
+	"github.com/docker/app/types/parameters"
+	"github.com/docker/cli/cli/command"
+	"github.com/docker/cli/cli/context/store"
+	cliopts "github.com/docker/cli/opts"
+	"github.com/pkg/errors"
+)
+
+func prepareCredentialSet(contextName string, contextStore store.Store, b *bundle.Bundle, namedCredentialsets []string) (map[string]string, error) {
+	creds := map[string]string{}
+	for _, file := range namedCredentialsets {
+		if _, err := os.Stat(file); err != nil {
+			file = filepath.Join(duffleHome().Credentials(), file+".yaml")
+		}
+		c, err := credentials.Load(file)
+		if err != nil {
+			return nil, err
+		}
+		values, err := c.Resolve()
+		if err != nil {
+			return nil, err
+		}
+		for k, v := range values {
+			if _, ok := creds[k]; ok {
+				return nil, fmt.Errorf("ambiguous credential resolution: %q is present in multiple credential sets", k)
+			}
+			creds[k] = v
+		}
+	}
+	if contextName != "" {
+		data, err := ioutil.ReadAll(store.Export(contextName, contextStore))
+		if err != nil {
+			return nil, err
+		}
+		creds["docker.context"] = string(data)
+	}
+	_, requiresDockerContext := b.Credentials["docker.context"]
+	_, hasDockerContext := creds["docker.context"]
+	if requiresDockerContext && !hasDockerContext {
+		return nil, errors.New("no target context specified. Use --target-context= or DOCKER_TARGET_CONTEXT= to define it")
+	}
+	return creds, nil
+}
+
+func getTargetContext(optstargetContext, currentContext string) string {
+	var targetContext string
+	switch {
+	case optstargetContext != "":
+		targetContext = optstargetContext
+	case os.Getenv("DOCKER_TARGET_CONTEXT") != "":
+		targetContext = os.Getenv("DOCKER_TARGET_CONTEXT")
+	}
+	if targetContext == "" {
+		targetContext = currentContext
+	}
+	return targetContext
+}
+
+func duffleHome() home.Home {
+	return home.Home(home.DefaultHome())
+}
+
+// prepareDriver prepares a driver per the user's request.
+func prepareDriver(dockerCli command.Cli) (driver.Driver, error) {
+	driverImpl, err := driver.Lookup("docker")
+	if err != nil {
+		return driverImpl, err
+	}
+	if d, ok := driverImpl.(*driver.DockerDriver); ok {
+		d.SetDockerCli(dockerCli)
+	}
+
+	// Load any driver-specific config out of the environment.
+	if configurable, ok := driverImpl.(driver.Configurable); ok {
+		driverCfg := map[string]string{}
+		for env := range configurable.Config() {
+			driverCfg[env] = os.Getenv(env)
+		}
+		configurable.SetConfig(driverCfg)
+	}
+
+	return driverImpl, err
+}
+
+func applyParameterValues(parameterValues map[string]string, parameterDefinitions map[string]bundle.ParameterDefinition, finalValues map[string]interface{}) error {
```

Comment:
I feel like all the parameter related functions have confusing naming (we added each of them separately in the original CNAB spike, and now that I see all of them together, it smells a little :D)

Reply:
Let's refactor that in a follow-up!